### PR TITLE
Cog refactor

### PR DESCRIPTION
**Pull request by Andrewki44**
Cogs reviewed and commands converted to slash commands.

`Admin` & `Trueskill` cogs added

To Do:
- Move as much out of `commands` as possible, into organised groups & cogs

### CODE REVIEW

**Comment by Andrewki44** [comment]
`common`, `list` & `random` cogs added.
Some reorganization of commands to `list.py`.

**Comment by Andrewki44** [comment]
I believe this is now in the final state for now.
Some of the most common / complicated commands have not been moved yet (`!add`, `!del`, `!sub` & `!autosub`) due to concerns over variable arguments in interactions, and not wanting to hold up everything else in the meantime.

I've tested fairly well as I've gone, and will do some more too.

I have maintained functionality in commands as-is as much as possible, but have made some minor changes necessary to convert to app_commands, as well as implementing some new checks, reorganizing cogs & utils, etc.

Appreciate any & all feedback

File: discord_bots/cogs/common.py
Comment (mcoot) [inline discord_bots/cogs/common.py:561]: I see some of these e.g. queue_autocomplete are defined identically in multiple cogs (this one doesn't appear to have any references). Could these be combined to a central place?

**Review by mcoot — COMMENTED** [review]
Broadly looks good to me. I can't claim to have validated the logic of every cog, but I looked at the first couple and the structure makes sense. A few small questions, but broadly speaking I think it's good to go. Would be good to get at least an additional reviewer before merge.

File: discord_bots/cogs/admin.py
Comment (Andrewki44) [reply discord_bots/cogs/admin.py:248]: It is, but I converted the command over as-is.
Outside of the scope of this PR.

I did take a quick look at what was involved with the dockerized pgsql, but was more than I was willing to dive into as a part of this.

File: discord_bots/cogs/admin.py
Comment (Andrewki44) [reply discord_bots/cogs/admin.py:445]: For the grouping, there was several ways it could've been done, but I decided on just having it be alphabetical in most circumstances. 
I think `utils.py` might be the only exception to that.

Further subdivision / grouping can certainly be looked at down the road, but as it is, none of these cogs exceed the 25 command maximum.

File: discord_bots/cogs/common.py
Comment (Andrewki44) [reply discord_bots/cogs/common.py:561]: Good catch, removed the unused autocomplete function from `common.py`.

You're correct that these are replicated a couple times, and 95% of them are identical. 
My only concern with externalising them from each individual cog was the decorator requirement.
As the autocomplete function is decorated with the slash command function (and so the slash command must be in scope), and not the other way around, it seemed tricky to move them elsewhere.

Creating a new `AutocompleteCommands` cog (or similar) doesn't compile, as you cannot use the function from another cog as the decorator inside the new cog.
Similarly, creating a new file (that is not a cog), and just loading the defs into there (similar to `commands` or `utils` compiles, but does not resolve the results when running the slash command it is attempting to autocomplete.

TLDR;
The autocomplete function seems to need to be in the same scope as the function it is referring to.

Comment (Andrewki44) [reply discord_bots/cogs/common.py]: The specific def referenced is removed in d4871a3

File: discord_bots/checks.py
Comment (Andrewki44) [reply discord_bots/checks.py:115]: Separating the conditional clauses removed a typing error that was present.
As if you try to interaction with `interaction.channel` before determining that it is `not None`, it presents an error.

Having a code path that returns out to False before checking if the `interaction.channel.id == CHANNEL_ID` removes the error from presenting.

Realistically, yes it could likely be compressed a bit, but I figured going for syntactic accuracy was preferable.

Comment (mcoot) [reply discord_bots/checks.py:115]: Maybe the typechecker can't narrow it but the `not interaction.channel` being first in the condition should resolve that issue since it short-circuits?

Comment (Andrewki44) [reply discord_bots/checks.py]: You make a fair point.
I've simplified the check in [acb7a20](https://github.com/tribesonecommunity/discord-bots/pull/93/commits/acb7a204e4d05402c361bc73c41b4638d6e35efa), and tested it as successful.

File: discord_bots/cogs/admin.py
Comment (StephenBartos) [inline discord_bots/cogs/admin.py:139]: `.first()` can definitely be used here instead to simplify the logic, but since it was like this before you can just add a TODO. Up to you if you want to refactor it or not, because the logic that follows assumes the correct player is `players[0]`

File: discord_bots/cogs/admin.py
Comment (StephenBartos) [inline discord_bots/cogs/admin.py:210]: the `app_commands.guild_only` decorator can get rid of this check, if you want to simplify it a bit. Just `assert guild` at the top if you do that

File: discord_bots/cogs/admin.py
Comment (StephenBartos) [inline discord_bots/cogs/admin.py:391]: does `outcome` need to be a string too or can it also be a list of `Literals`?

File: discord_bots/cogs/admin.py
Comment (StephenBartos) [inline discord_bots/cogs/admin.py:396]: For the future, we may want to move away from the BE/DS convention here, and just do `team0, team1` or `team1, team2`.

File: discord_bots/cogs/admin.py
Comment (StephenBartos) [reply discord_bots/cogs/admin.py:445]: As we've discussed over discord, further command grouping can come down with more general refactors to the _logic_ of the commands. I don't really care about the ordering within the files, since the darker doesn't do it for you and you can always use an IDE or grep to find whatever one you want. Of course that would be nice to have, but unless there's some automated way to do it via the pre-commit hook, then I don't want anyone to have to do that themselves by hand. Alphabetical is preferred

File: discord_bots/cogs/admin.py
Comment (StephenBartos) [inline discord_bots/cogs/admin.py:448]: same thing as before, where `.first()` can be used

File: discord_bots/cogs/admin.py
Comment (StephenBartos) [inline discord_bots/cogs/admin.py:472]: Use `app_commands.guild_only` to simplify

File: discord_bots/cogs/admin.py
Comment (StephenBartos) [inline discord_bots/cogs/admin.py:542]: the db specific commands don't work with postgres anymore, so it should be marked clearly both with a comment and an early return inside of the function that tells the user so.

File: discord_bots/cogs/admin.py
Comment (StephenBartos) [inline discord_bots/cogs/admin.py:577]: This might be dangerous to keep around with the dockerized setup, since any admin can use it as well. I'm not sure how it behaves with docker, though

File: discord_bots/cogs/admin.py
Comment (StephenBartos) [inline discord_bots/cogs/admin.py:590]: Since these are "meme" functions, we may just want to leave them as prefix command or (if there's a way to) hide them. Not really sure, since they just pollute the command list

File: discord_bots/cogs/admin.py
Comment (StephenBartos) [inline discord_bots/cogs/admin.py:654]: same as before: use `.first()`

**Review by StephenBartos — COMMENTED** [review]
Reviewed checks.py and admin.py:

I understand that you migrated these commands over as-is, so I don't expect you to address every single comment with a code change; for nearly all of these, just adding a `TODO` or some other comment is good enough for a second pass in the future. I'm just taking this chance to review each of these commands. Awesome work so far!

- [x] checks.py
- [x] admin.py
- [x] base.py
- [ ] category.py
- [ ] common.py
- [ ] economy.py
- [ ] in_progress_game.py
- [ ] list.py
- [ ] map.py
- [ ] notification.py
- [ ] player.py
- [ ] queue.py
- [ ] raffle.py
- [ ] random.py
- [ ] rotation.py
- [ ] schedule.py
- [ ] trueskill.py
- [ ] vote.py
- [ ] commands.py
- [ ] config.py
- [ ] main.py
- [ ] models.py
- [ ] utils.py

File: discord_bots/checks.py
Comment (Andrewki44) [reply discord_bots/checks.py:25]: Updated in next commit

File: discord_bots/cogs/admin.py
Comment (Andrewki44) [reply discord_bots/cogs/admin.py:97]: Updated in next commit

File: discord_bots/cogs/admin.py
Comment (Andrewki44) [reply discord_bots/cogs/admin.py:117]: Updated in next commit

File: discord_bots/cogs/admin.py
Comment (Andrewki44) [reply discord_bots/cogs/admin.py:139]: Replaced `players` with `player: Player | None`, and modified function logic to understand the singular nature of `.first()` as opposed to `.all()`

File: discord_bots/cogs/admin.py
Comment (Andrewki44) [reply discord_bots/cogs/admin.py:210]: Added `@app_commands.guild_only()`, and simplified function logic to remove the `if interaction.guild`

File: discord_bots/cogs/admin.py
Comment (Andrewki44) [reply discord_bots/cogs/admin.py:391]: Not sure if I'm misunderstanding, but `outcome` already is a Literal list of of strings.
`outcome: Literal["Tie", "BE", "DS"]`

File: discord_bots/cogs/admin.py
Comment (Andrewki44) [reply discord_bots/cogs/admin.py:396]: Added a `TODO`

File: discord_bots/cogs/admin.py
Comment (StephenBartos) [reply discord_bots/cogs/admin.py:391]: I think I misunderstood myself, since this is just for the description. I don't think outcome even needs a description, since it will be autofilled to 3 valid options: "Tie", "BE", or "DS" -- my point was that this doesn't match precisely

Comment (Andrewki44) [reply discord_bots/cogs/admin.py:391]: Understood.
I do prefer to have the `.describe()` there for it anyway, just for completeness sake. I have updated it thought to match the `Literal` (`"<Tie|BE|DS>"`)

File: discord_bots/cogs/admin.py
Comment (Andrewki44) [reply discord_bots/cogs/admin.py:448]: Changed to `.first()`, and subsequent logic updated.

File: discord_bots/cogs/admin.py
Comment (Andrewki44) [reply discord_bots/cogs/admin.py:391]: Updated description to `"Tie, BE, DS"` as discussed in Discord

File: discord_bots/cogs/admin.py
Comment (Andrewki44) [reply discord_bots/cogs/admin.py:472]: added `@app_commands.guild_only()`, and modified subsequent logic to match.

File: discord_bots/cogs/admin.py
Comment (Andrewki44) [reply discord_bots/cogs/admin.py:542]: Added check on `config.DATABASE_URI` to return early if setup for postgres.
Added comments & `TODO`.

Similarly updated `createdbbackup()`

File: discord_bots/cogs/admin.py
Comment (Andrewki44) [reply discord_bots/cogs/admin.py:577]: The early return on postgres db's will be sufficient for now to block any bad interactions. In future this & `createdbbackup()` should be updated to support the dockerized db.
`TODO` added

File: discord_bots/cogs/admin.py
Comment (Andrewki44) [reply discord_bots/cogs/admin.py:590]: I would rather not leave anything as prefix / context commands, as the direction from discord itself is to move away from context commands entirely.
https://support-dev.discord.com/hc/en-us/articles/4404772028055-Message-Content-Privileged-Intent-FAQ

File: discord_bots/cogs/admin.py
Comment (Andrewki44) [reply discord_bots/cogs/admin.py:654]: Updated to use `.first()`, along with the subsequent logic.

**Comment by Andrewki44** [comment]
`player.leaderboard_enabled` was not being considered in `print_leaderboard()`.
Fixed in ffbb2f5

File: discord_bots/cogs/admin.py
Comment (StephenBartos) [reply discord_bots/cogs/admin.py:615]: If the prefix command @mentions the bot or is used inside of the bot's DM, then it doesn't required the message content privileged intent.

File: discord_bots/cogs/category.py
Comment (StephenBartos) [inline discord_bots/cogs/category.py:276]: `app_copmmands.rename` `category_name` to `category`

File: discord_bots/cogs/common.py
Comment (StephenBartos) [inline discord_bots/cogs/common.py:366]: Please use a block comment instead haha, I hope you didn't add each # by hand
Eg:
```python
"""
Multiline 
block 
comment
"""
```

File: discord_bots/cogs/economy.py
Comment (StephenBartos) [inline discord_bots/cogs/economy.py:1]: I'm going to skip over reviewing economy, since you're way more familiar with the file and I trust your judgment here with respect to these changes

File: discord_bots/cogs/in_progress_game.py
Comment (StephenBartos) [inline discord_bots/cogs/in_progress_game.py:190]: Good catch, since this can be confused with the internal view.is_finished field used by discord.py

File: discord_bots/cogs/in_progress_game.py
Comment (StephenBartos) [inline discord_bots/cogs/in_progress_game.py:407]: Sorry, I got confused by the diff and thought you changed this above. Rename this to `is_game_finished`

File: discord_bots/cogs/in_progress_game.py
Comment (StephenBartos) [inline discord_bots/cogs/in_progress_game.py:668]: This can be simplified to `if isinstance(game_history_channel, TextChannel):`, since `None` is a type and will fail in the `isinstance` check

File: discord_bots/cogs/map.py
Comment (StephenBartos) [inline discord_bots/cogs/map.py:46]: Instead of outright renaming "map_short_name" to "map_name", you could have used the app_commans.rename decorator. This was probably written before you knew about it or something, but the refactor is fine.

File: discord_bots/cogs/notification.py
Comment (StephenBartos) [inline discord_bots/cogs/notification.py:14]: This should be `class NotificationCommands`

File: discord_bots/cogs/trueskill.py
Comment (StephenBartos) [inline discord_bots/cogs/trueskill.py:205]: block comment

File: discord_bots/utils.py
Comment (StephenBartos) [inline discord_bots/utils.py:68]: use underscore_case like the other functions. Maybe something  like `build_category_str` or `get_category_str`. Whatever you want

File: discord_bots/utils.py
Comment (StephenBartos) [inline discord_bots/utils.py:1484]: can remove my comment here

**Review by StephenBartos — APPROVED** [review]
Ready to merge, except for some very _very_ minor feedback. Really awesome work cleaning up all of the commands and reorganizing everything. It definitely seemed like a huge, tedious task but I really appreciate it. This is an enormous step forwards in cleaning everything up. Thank you for your patience as well!

File: discord_bots/cogs/category.py
Comment (Andrewki44) [reply discord_bots/cogs/category.py:26]: Good thinking, have moved it across to `/queue clearcategory <queue_name>`
It is a queue related command, but the intention is to remove whatever category is assigned to the given queue. 
Therefore the queue is the input, not the category.
Added it to the autocomplete as well.

File: discord_bots/cogs/category.py
Comment (Andrewki44) [reply discord_bots/cogs/category.py:56]: If you're looking at my dev bot, that'll be because it's synced to my `RotationConfigure` fork at the moment, which also has `CategoryConfigure` & `MapConfigure` within it.

As a part of `CategoryConfigure` this command was removed in favour of `/category configure <category_name>`.
That is not the case for this PR though, and I have tested this successful.

File: discord_bots/cogs/common.py
Comment (Andrewki44) [reply discord_bots/cogs/common.py:93]: Added a rename decorator.
This could be done more-so around the place, some more general clean up to be done separately.

File: discord_bots/cogs/category.py
Comment (Andrewki44) [reply discord_bots/cogs/category.py:276]: Added a rename decorator.
Did so for the remaining functions in `category.py`

File: discord_bots/cogs/common.py
Comment (Andrewki44) [reply discord_bots/cogs/common.py:366]: I just used the `Ctrl+/` shortcut in VS Code to be honest, and that's how it comes out.

Swapped over to docstring instead, which is convenient cause VS Code will let me collapse that now.

File: discord_bots/cogs/in_progress_game.py
Comment (Andrewki44) [reply discord_bots/cogs/in_progress_game.py:407]: Renamed

File: discord_bots/cogs/in_progress_game.py
Comment (Andrewki44) [reply discord_bots/cogs/in_progress_game.py:668]: Done

File: discord_bots/cogs/notification.py
Comment (Andrewki44) [reply discord_bots/cogs/notification.py:14]: Good pickup. I was copying the class structure from file to file and must have missed the rename on this one, thanks!
Weird that it didn't give any kind of error though, I guess cause on sync it's syncing the command tree via the unique group.

File: discord_bots/cogs/trueskill.py
Comment (Andrewki44) [reply discord_bots/cogs/trueskill.py:205]: Done

File: discord_bots/utils.py
Comment (Andrewki44) [reply discord_bots/utils.py:68]: I believe this was added in a PR after this was originally raised, so was likely missed from my review.
Good catch, have renamed it here, and in all references.

File: discord_bots/utils.py
Comment (Andrewki44) [reply discord_bots/utils.py:1484]: Done

**Comment by Andrewki44** [comment]
@StephenBartos all comments have been reviewed and updated. 
Should be ready to merge.

**Pull request merged**
